diff --git a/node.go b/node.go
index e86c0c3..028867c 100644
--- a/node.go
+++ b/node.go
@@ -48,7 +48,7 @@ type Node struct {
 
 // InnerText returns the text between the start and end tags of the object.
 func (n *Node) InnerText() string {
-	var output func(*bytes.Buffer, *Node)
+	/*var output func(*bytes.Buffer, *Node)
 	output = func(buf *bytes.Buffer, n *Node) {
 		switch n.Type {
 		case TextNode:
@@ -64,7 +64,18 @@ func (n *Node) InnerText() string {
 
 	var buf bytes.Buffer
 	output(&buf, n)
-	return buf.String()
+	return buf.String()*/
+
+	if (n.Type == TextNode) {
+		return n.Data
+	} else if (n.Type == ElementNode) &&
+	(n.FirstChild != nil) &&
+	(n.FirstChild.Type == TextNode) {
+		return n.FirstChild.Data
+	}
+
+
+	return ""
 }
 
 func (n *Node) sanitizedData(preserveSpaces bool) string {
diff --git a/query.go b/query.go
index 146c2a4..575fedd 100644
--- a/query.go
+++ b/query.go
@@ -49,6 +49,17 @@ func CreateXPathNavigator(top *Node) *NodeNavigator {
 	return &NodeNavigator{curr: top, root: top, attr: -1}
 }
 
+//Evaluate XPath expression, the expression should evaluate to true or false
+func Eval(top, ctx *Node, exp *xpath.Expr) bool {
+	if exp == nil {
+		return false
+	}
+
+	v := exp.Evaluate(&NodeNavigator{curr: ctx, ctxt: ctx, root: top, attr: -1})
+
+	return v.(bool)
+}
+
 func getCurrentNode(it *xpath.NodeIterator) *Node {
 	n := it.Current().(*NodeNavigator)
 	if n.NodeType() == xpath.AttributeNode {
@@ -145,7 +156,7 @@ func FindEachWithBreak(top *Node, expr string, cb func(int, *Node) bool) {
 }
 
 type NodeNavigator struct {
-	root, curr *Node
+	root, curr, ctxt *Node
 	attr       int
 }
 
@@ -212,6 +223,17 @@ func (x *NodeNavigator) MoveToRoot() {
 	x.curr = x.root
 }
 
+func (x *NodeNavigator) MoveToContext() {
+	x.curr = x.ctxt
+}
+
+func (x *NodeNavigator) CurrentPrefix() string {
+	if (x.ctxt != nil) {
+		return x.ctxt.Prefix
+	}
+	return ""
+}
+
 func (x *NodeNavigator) MoveToParent() bool {
 	if x.attr != -1 {
 		x.attr = -1
