diff --git a/plugins/clish/sym_script.c b/plugins/clish/sym_script.c
index 93cc88f..7dee09b 100644
--- a/plugins/clish/sym_script.c
+++ b/plugins/clish/sym_script.c
@@ -37,6 +37,10 @@ CLISH_PLUGIN_OSYM(clish_script)
 	if (!script) /* Nothing to do */
 		return 0;
 
+	char *cmd = clish_shell__get_full_line(clish_context);
+	setenv("USER_COMMAND", cmd, 1);
+	lub_string_free(cmd);
+
 	/* Find out shebang */
 	if (action)
 		shebang = clish_action__get_shebang(action);
@@ -80,6 +84,12 @@ CLISH_PLUGIN_OSYM(clish_script)
 	lub_string_cat(&command, shebang);
 	lub_string_cat(&command, " ");
 	lub_string_cat(&command, fifo_name);
+	lub_string_cat(&command, " 2> /dev/stdout");
+
+	sigset_t sigs;
+	sigemptyset(&sigs);
+	sigaddset(&sigs, SIGINT);
+	sigprocmask(SIG_UNBLOCK, &sigs, NULL);
 
 	res = system(command);
 
